src/wp-includes/class-wp-theme-json.php

```diff
@@ -3236,17 +3236,19 @@ public function merge( $incoming ) {
 		 * some values provide exceptions, namely style values that are
 		 * objects and represent unique definitions for the style.
 		 */
-		$style_nodes = static::get_styles_block_nodes();
-		foreach ( $style_nodes as $style_node ) {
-			$path = $style_node['path'];
-			/*
-			 * Background image styles should be replaced, not merged,
-			 * as they themselves are specific object definitions for the style.
-			 */
-			$background_image_path = array_merge( $path, static::PROPERTIES_METADATA['background-image'] );
-			$content               = _wp_array_get( $incoming_data, $background_image_path, null );
-			if ( isset( $content ) ) {
-				_wp_array_set( $this->theme_json, $background_image_path, $content );
+		if ( isset( $incoming_data['styles']['blocks'] ) && is_array( $incoming_data['styles']['blocks'] ) ) {
+			foreach ( $incoming_data['styles']['blocks'] as $block_name => $block_styles ) {
+				/*
+				 * Background image styles should be replaced, not merged,
+				 * as they themselves are specific object definitions for the style.
+				 */
+				if ( isset( $block_styles['background']['backgroundImage'] ) ) {
+					$background_image_path = array_merge( array( 'styles', 'blocks', $block_name ), static::PROPERTIES_METADATA['background-image'] );
+					$content               = _wp_array_get( $incoming_data, $background_image_path, null );
+					if ( isset( $content ) ) {
+						_wp_array_set( $this->theme_json, $background_image_path, $content );
+					}
+				}
 			}
```

Comment:
I believe there is one crucial piece missing here, which is to account for specific element styles. Please check the `get_block_nodes()` method (called by `get_styles_blocks_nodes()`: You will see that it also creates nodes for any relevant `$theme_json['styles']['blocks'][ $name ]['elements']` (which here would be `$incoming_data['styles']['blocks'][ $name ]['elements']`).

How about we introduce a new private helper method like `get_block_node_paths( $theme_json )`, which returns just the list of all node paths? That way you can keep this method almost like it was before, but it calls a cheaper method.

Reply:
Thanks for the feedback. In https://github.com/WordPress/wordpress-develop/pull/7486/commits/46640a7e77b5e6f718203af373fce4c0ac2c021b I Introduce new function.